### PR TITLE
Middleware article and sample project.

### PR DESCRIPTION
**Pull request by ardalis**
Fixes issue #109

### CODE REVIEW

**Comment by danroth27** [comment]
@Tratcher Please review

File: aspnet/fundamentals/middleware.rst [hunk 1]
```diff
@@ -1,8 +1,176 @@
-.. include:: /../common/stub-topic.txt
+Middleware
+==========
+By `Steve Smith`_
 
-|stub-icon| Middleware
-======================
+:doc:`owin`, the Open Web Interface for .NET, describes small application components that can be incorporated into an HTTP request pipeline as middleware. ASP.NET 5 has integrated support for middleware, which are wired up in an application's ``Configure`` method during :doc:`startup`.
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: Remove references to OWIN.

File: aspnet/fundamentals/middleware.rst [hunk 2]
```diff
@@ -1,8 +1,176 @@
...
+It's important to realize that request delegate, as written here, will terminate the pipeline, regardless of other calls to ``App.Run`` that you may include. In the following example, only the first delegate ("Hello, World!") will be executed and displayed.
+
+.. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
+	:language: c#
+	:linenos:
+	:lines: 20-31
+	:emphasize-lines: 5
+	:dedent: 8
+
+You chain multiple request delegates together making a different call, with a ``next`` parameter representing the next delegate in the pipeline. Note that just because you're calling it "next" doesn't mean you can't perform actions both before and after the next delegate, as this example demonstrates:
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: But beware of trying to modify the HttpResponse after calling next, as one of the following components most likely wrote to the response and caused it to be sent to the client.

File: aspnet/fundamentals/middleware.rst [hunk 3]
```diff
@@ -1,8 +1,176 @@
...
+
+.. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
+	:language: c#
+	:linenos:
+	:lines: 34-49
+	:emphasize-lines: 5,8,14
+	:dedent: 8
+
+.. note:: This ``ConfigureLogInline`` method is called when the application is run with an environment set to ``LogInline``. Learn more about :doc:`environments`. We will be using variations of ``Configure[Environment]`` to show different options in the rest of this article.
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: Can you link to the Startup article here?

File: aspnet/fundamentals/middleware.rst [hunk 4]
```diff
@@ -1,8 +1,176 @@
...
+
+In the above example, the call to ``await next.Invoke()`` will call into the delegate on line 14. The client will receive the expected response ("Hello from LogInline"), and the server's console output includes both the before and after messages, as you can see here:
+
+.. image:: middleware/_static/console-loginline.png
+
+Run, Map, and Use
+^^^^^^^^^^^^^^^^^
+
+You configure the HTTP pipeline using the `extensions <https://github.com/aspnet/HttpAbstractions/tree/dev/src/Microsoft.AspNet.Http.Abstractions/Extensions>`_ ``Run``, ``Map``, and ``Use``. By convention, the ``Run`` method is simply a shorthand way of adding middleware to the pipeline that doesn't call any other middleware (that is, it will not call a ``next`` request delegate). The following two examples (one using ``Run`` and the other ``Use``) are equivalent to one another, since the second one doesn't use its ``next`` parameter:
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: Thus Run should only be used at the end of your pipeline.

Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: Note that Run is a convention, not just a single extension. Run[Middleware] extensions may be provided by some components that should only run at the end of the pipeline.

File: aspnet/fundamentals/middleware.rst [hunk 5]
```diff
@@ -1,8 +1,176 @@
...
+.. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
+	:language: c#
+	:linenos:
+	:lines: 65-79
+	:emphasize-lines: 3,11
+	:dedent: 8
+
+.. note:: The `IApplicationBuilder interface <https://github.com/aspnet/HttpAbstractions/blob/1.0.0-beta5/src/Microsoft.AspNet.Http.Abstractions/IApplicationBuilder.cs#L17>`_ itself exposes a single ``Use`` method, so technically they're not all *extension* methods.
+
+We've already seen several examples of how to build a request pipeline with ``Use``. The ``Map`` extension method is used to match request delegates based on a request's path. ``Map`` simply accepts a path and a function that configures a separate middleware pipeline. In this example, any request with the base path of ``/maptest`` will be handled by the pipeline configured in the ``HandleMapTest`` method.
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: Map* is used as a convention for branching/forking the pipeline. Current implementations are Map(Path) and MapWhen.

Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: Note that for Map(Path) the matched path segment(s) are removed from HttpRequest.Path and appended to HttpReqeust.PathBase on each request.

File: aspnet/fundamentals/middleware.rst [hunk 6]
```diff
@@ -1,8 +1,176 @@
...
+	*  - :doc:`Authentication </security/sociallogins>`
+	   - Provides authentication support.
+	*  - :doc:`diagnostics`
+	   - Includes support for error pages and runtime information.
+	*  - :doc:`routing`
+	   - Define and constrain request routes.
+	*  - :doc:`Static Files <static-files>`
+	   - Provides support for serving static files, and directory browsing.
+
+.. Do we need to list any of these? MVC, AuthorizeBasicMiddleware, Session, CORS, ResponseCaching
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: List Session and CORS.

File: aspnet/fundamentals/middleware.rst [hunk 7]
```diff
@@ -1,8 +1,176 @@
...
+
+For more complex request handling functionality, the ASP.NET team recommends implementing the middleware in its own class, and exposing an ``IApplicationBuilder`` extension method that can be called from the ``Configure`` method. The simple logging middleware shown in the previous example can be converted into a middleware class that takes in the next ``RequestDelegate`` in its constructor and supports an ``Invoke`` method as shown:
+
+.. literalinclude:: middleware/sample/src/MiddlewareSample/RequestLoggerMiddleware.cs
+	:language: c#
+	:caption: RequestLoggerMiddleware.cs
+	:linenos:
+	:emphasize-lines: 13, 19
+
+The middleware follows the `Explicit Dependencies Principle <http://deviq.com/explicit-dependencies-principle/>`_ and exposes all of its dependencies in its constructor. The extension method is responsible for providing the required dependencies. You can create overloads of the extension methods that allow dependencies to be passed in from the ``Configure`` method, as well. The ``UseRequestLogger`` extension method is shown below:
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: Middleware can take advantage of the `UseMiddleware<T>` extension to inject DI services directly into their constructors. See:
https://github.com/aspnet/StaticFiles/blob/1.0.0-beta5/src/Microsoft.AspNet.StaticFiles/StaticFileExtensions.cs#L44
https://github.com/aspnet/StaticFiles/blob/1.0.0-beta5/src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs#L30
Any non-DI parameters can also be passed in so long as they come after the DI parameters in the constructor.

**Comment by Tratcher** [comment]
:watch:

File: aspnet/fundamentals/middleware.rst (hunk 1 above)
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: +1 ASP.NET 5 middleware really has nothing to do with OWIN.

File: aspnet/fundamentals/middleware.rst [hunk 8]
```diff
@@ -1,8 +1,176 @@
...
+	- `Built-in middleware`_
+	- `Writing middleware`_
 
-.. _issue: https://github.com/aspnet/Docs/issues/109
+`Download sample from GitHub <https://github.com/aspnet/docs/aspnet/fundamentals/middleware/_static/sample>`_. 
+
+What is middleware
+------------------
+
+The `OWIN specification <http://owin.org/spec/spec/owin-1.0.0.html#Definition>`_ defines middleware as:
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: The middleware model in ASP.NET 5 really is completely independent of OWIN. We support OWIN through an adapter layer. The core middleware model in ASP.NET 5 isn't based on OWIN at all.

**Comment by danroth27** [comment]
Please update the sample to beta6 before merging.

File: aspnet/fundamentals/middleware.rst (hunk 1 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Done.

File: aspnet/fundamentals/middleware.rst (hunk 2 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Noted.

File: aspnet/fundamentals/middleware.rst (hunk 3 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Done.

File: aspnet/fundamentals/middleware.rst (hunk 4 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Noted.

File: aspnet/fundamentals/middleware.rst (hunk 5 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Noted.

File: aspnet/fundamentals/middleware.rst (hunk 6 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Done.

File: aspnet/fundamentals/middleware.rst (hunk 8 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Updated to remove references to OWIN spec.

File: aspnet/fundamentals/middleware.rst (hunk 7 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Noted.

**Comment by ardalis** [comment]
Should be about ready to merge. /cc @danroth27 @Tratcher

File: aspnet/fundamentals/middleware.rst [hunk 9]
```diff
@@ -1,8 +1,180 @@
...
+	- `Built-in middleware`_
+	- `Writing middleware`_
 
-.. _issue: https://github.com/aspnet/Docs/issues/109
+`Download sample from GitHub <https://github.com/aspnet/docs/aspnet/fundamentals/middleware/_static/sample>`_. 
+
+What is middleware
+------------------
+
+Middleware are components that form a pipeline between a server and application for individual requests and responses. Each component can choose whether to pass the request on to the next component in the pipeline, and can perform certain actions before and after the next component in the pipeline. Request delegates are used to build this request pipeline, which are then used to handle each incoming HTTP request to your application.
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: "Middleware are components that are assembled into an application pipeline to handle requests and responses."

File: aspnet/fundamentals/middleware.rst [hunk 10]
```diff
@@ -1,8 +1,180 @@
...
+Summary
+-------
+
+Middleware provide simple components for adding features to individual web requests. Applications configure their request pipelines in accordance with the features they need to support, and thus have fine-grained control over the functionality each request uses. Developers can easily create their own middleware to provide additional functionality to ASP.NET applications.
+
+Additional Resources
+--------------------
+
+- :doc:`startup`
+- :doc:`owin`
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware.rst]: No need to reference OWIN

File: aspnet/fundamentals/middleware/sample/global.json [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿{
+  "projects": [ "src", "test" ],
+  "sdk": {
+    "version": "1.0.0-beta5",
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware/sample/global.json]: beta6? or just remove this

File: aspnet/fundamentals/middleware/sample/src/MiddlewareSample/RequestLoggerExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.Logging;
+
+namespace MiddlewareSample
+{
+    public static class RequestLoggerExtensions
+    {
+        public static IApplicationBuilder UseRequestLogger(this IApplicationBuilder builder)
+        {
+            var env = builder.ApplicationServices.GetService<IHostingEnvironment>();
```
Comment (Tratcher) [inline aspnet/fundamentals/middleware/sample/src/MiddlewareSample/RequestLoggerExtensions.cs]: See https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication/AuthenticationMiddleware.cs#L35 for our recommended way to set up logging. I wouldn't want anyone to copy this sample code.

**Comment by Tratcher** [comment]
:shipit: after comments.

File: aspnet/fundamentals/middleware/sample/global.json (hunk 11 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware/sample/global.json]: Removed

File: aspnet/fundamentals/middleware.rst (hunk 10 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Removed

File: aspnet/fundamentals/middleware.rst (hunk 9 above)
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Great, thanks.

**Pull request merged**